Sources/TensorFlow/Core/LazyTensorTrace.swift

```diff
@@ -0,0 +1,160 @@
+    }
+
+    private var lazyOpsCache: [ObjectIdentifier: LazyTensorOperation] = [:]
+
+    private func updateCacheAndOperations(
+        _ id: ObjectIdentifier, _ node: LazyTensorOperation) {
+        lazyOpsCache[id] = node
+        node.id = "\(operations.count)"
+        operations.append(node)
+    }
+
+    private func makeConstTensor(with conc: TFETensorHandle) -> LazyTensor {
+        let cTensorHandle = conc._cTensorHandle
+        let result = LazyTensorOperation("Const", 1)
+        let dtype = TensorDataType(TFE_TensorHandleDataType(cTensorHandle))
+        let dtypeAttr = LazyTensorOperation.Attribute.TensorDataTypeValue(dtype)
+        result.attrs = [
+            "dtype": dtypeAttr,
+            "value": LazyTensorOperation.Attribute.ConstTensor(conc)]
+        updateCacheAndOperations(ObjectIdentifier(conc), result)
+        return LazyTensor(_lazy: result, index: 0)
+    }
+
+    private func makePlaceholderTensor(
+        with conc: TFETensorHandle
+    ) -> LazyTensor {
+        let cTensorHandle = conc._cTensorHandle
+        let dtype = TensorDataType(TFE_TensorHandleDataType(cTensorHandle))
+        let dtypeAttr = LazyTensorOperation.Attribute.TensorDataTypeValue(dtype)
+        let placeholder = LazyTensorOperation("Placeholder", 1)
+        placeholder.attrs = ["dtype": dtypeAttr]
+        updateCacheAndOperations(ObjectIdentifier(conc), placeholder)
+        inputs.append(placeholder)
+        inputValues.append(conc)
+        return LazyTensor(_lazy: placeholder, index: 0)
+    }
+
+    private func makeConstTensorOrPlaceholder(
+        with conc: TFETensorHandle, asConst: Bool
+    ) -> LazyTensor {
+        let id = ObjectIdentifier(conc)
+        if let lazyOp = lazyOpsCache[id] {
+            return LazyTensor(_lazy: lazyOp, index: 0)
+        }
+        return asConst
+            ? makeConstTensor(with: conc)
+            : makePlaceholderTensor(with: conc)
+    }
+
+    // Return the original tensor or a concret tensor that is promoted to a
+    // placeholder input.
+    private func maybePromotedTensor(_ lazyHandle: LazyTensor) -> LazyTensor {
+        switch lazyHandle.handle {
+        case LazyTensor.Handle.concrete(let h, let materialized):
+            return makeConstTensorOrPlaceholder(
+                with: h, asConst: !materialized)
+        case LazyTensor.Handle.symbolic(let lazyOp, let index, _): do {
+                if let outputs = lazyOp.outputs {
+                    return makeConstTensorOrPlaceholder(
+                        with: outputs[index], asConst: false)
+                } else {
+                    return LazyTensor(
+                        _lazy: collectLazyOp(lazyOp), index: index)
+                }
+            }
+        }
+    }
+
+    private func maybePromotedInput(
+        _ input: LazyTensorOperation.Input
+    ) -> LazyTensorOperation.Input {
+        switch input {
+        case LazyTensorOperation.Input.single(let h):
```

Comment:
Omit type names when pattern matching enum cases.

```suggestion
        case .single(let h):
```